LibraBFT/Abstract/BFT.agda

```diff
@@ -6,11 +6,479 @@
+                        | intersectElem-∈ y∈xs sxs
+                        | +-∸-assoc (length xs) (s≤s (length-int-≤ xs ys sxs))
+                        | sym (+-∸-assoc (length xs) (length-int-≤ xs ys sxs))
+                        = union-length≡ sxs sys
+   ... | no  y∉xs rewrite unionElem-∉ (union-∉ (h∉t (y₁ ∷ sys)) y∉xs)
+                        | intersectElem-∉ y∉xs
+                        | +-suc (length xs) (length ys)
+                        | +-∸-assoc 1 (≤-stepsˡ (length xs) (length-int-≤ xs ys sxs))
+                        = cong suc (union-length≡ sxs sys)
+
+
+
+
+   m∸n≤o⇒m∸o≤n : ∀ (x z w : ℕ) → x ∸ z ≤ w → x ∸ w ≤ z
+   m∸n≤o⇒m∸o≤n x zero w p≤ rewrite m≤n⇒m∸n≡0 p≤ = z≤n
+   m∸n≤o⇒m∸o≤n zero (suc z) w p≤ rewrite 0∸n≡0 w = z≤n
+   m∸n≤o⇒m∸o≤n (suc x) (suc z) w p≤ = ≤-trans (∸-suc-≤ x w) (s≤s (m∸n≤o⇒m∸o≤n x z w p≤))
+     where  ∸-suc-≤ : ∀ (x w : ℕ) → suc x ∸ w ≤ suc (x ∸ w)
+            ∸-suc-≤ x zero = ≤-refl
+            ∸-suc-≤ zero (suc w) rewrite 0∸n≡0 w = z≤n
+            ∸-suc-≤ (suc x) (suc w) = ∸-suc-≤ x w
+
+
+
+
+   quorumInt>biz : ∀ (xs ys : List Member)
+                    → QSize ≤ length xs
+                    → QSize ≤ length ys
+                    → length xs + length ys ∸ authorsN ≤ length (intersect xs ys)
+                    → bizF + 1 ≤ length (intersect xs ys)
+   quorumInt>biz xs ys q≤x q≤y ≤int =
+     let p₁ = ≤-trans (∸-monoˡ-≤ authorsN (+-mono-≤ q≤x q≤y)) ≤int
+         p₂ = subst (_≤ length (intersect xs ys)) (simpExp₁ authorsN bizF) p₁
+         p₃ = ≤-trans (∸-monoˡ-≤ (2 * bizF) isBFT) p₂
+     in subst (_≤ length (intersect xs ys)) (simpExp₂ bizF) p₃
+       where  simpExp₁ : ∀ (x y : ℕ) → (x ∸ y) + (x ∸ y) ∸ x ≡ x ∸ (2 * y)
+              simpExp₁ x y rewrite sym (*-identityʳ (x ∸ y))
+                                 | sym (*-distribˡ-+ (x ∸ y) 1 1)
+                                 | *-comm (x ∸ y) 2
+                                 | *-distribˡ-∸ 2 x y
+                                 | ∸-+-assoc (2 * x) (2 * y) x
+                                 | +-comm (2 * y) x
+                                 | sym (∸-+-assoc (2 * x) x (2 * y))
+                                 | +-identityʳ x
+                                 | m+n∸n≡m x x = refl
+
+              simpExp₂ : ∀ (x : ℕ) → suc (3 * x) ∸ 2 * x ≡ x + 1
+              simpExp₂ x rewrite +-∸-assoc 1 (*-monoˡ-≤ x {2} {3} (s≤s (s≤s z≤n)))
+                               | sym (*-distribʳ-∸ x 3 2)
+                               | sym (+-suc x 0) = refl
+
+
+
+   span-hon : ∀ {xs dis hon : List Member} {x : Member}
+              → span Meta-dishonest? xs ≡ (dis , x ∷ hon)
+              → x ∈ xs ×  Meta-Honest-PK (getPubKey x)
+   span-hon {x₁ ∷ []} {dis} {hon} {x} eq
+     with Meta-dishonest? x₁
+   span-hon {x₁ ∷ []} {dis} {hon} {x} () | yes prf
+   span-hon {x₁ ∷ []} {[]} {[]} {x₁} refl | no ¬dis = here refl , ¬dis
+   span-hon {x₁ ∷ xs} {dis} {hon} {x} eq
```

Comment:
There is a catchall warning here, because there is no reason why `xs` can't be `[]`, which is covered by the previous cases as well.  Generally, we like to avoid warnings unless there is a good (documented) reason, or at least a `TODO-n` comment with some explanation of why it's there, so hopefully someone can address it later.

However, looking closer at this proof to see if I could easily remove the warning, I found that the proof is not as readable as most of your proofs.  I made another proof that doesn't have any warnings, and is more readable (to me, at least), though it's not perfect.  I'm pasting it below, feel free to take it or modify it if you like.

A few things I did here:
* Factored out a generic version of the lemma I needed, which makes it less cluttered with application-specific noise, and potentially enables reuse in other contexts.  I think there are a number of your proofs that would benefit from similar treatment.  Sufficiently generic lemmas should go in `Lemmas.agda`.
* Made the induction mirror that of the definition of `span`.   I think this is a big part of the reason it's simpler.

```
   -- Surely there is a library property?  If not, this should go in Lemmas
   []≢a∷as : ∀ {ℓ : Level} {A : Set ℓ}{a : A}{as1 as2 : List A}
           → as1 ≡ [] → as1 ≡ a ∷ as2
           → ⊥
   []≢a∷as {as1 = []} refl ()

   span-correct-1 : ∀ {ℓ₁ ℓ₂ A}{P : Unary-Pred {ℓ₁} A ℓ₂}{P? : Unary-Decidable P}{xs rest : List A}{x}
                  → proj₂ (span P? xs) ≡ x ∷ rest
                  → x ∈ xs × ¬ (P x)
   span-correct-1 {P? = P?} {[]} ()
   span-correct-1 {P? = P?} {xs = x ∷ xs} eq
     with does (P? x) | Dec.proof (P? x)
   ...| false         | ofⁿ ¬p rewrite ∷-injectiveˡ eq = here refl , ¬p
   ...| true          | _
      with span P? xs | inspect (span P?) xs
   ...| (_ , [])      | _ = ⊥-elim ([]≢a∷as refl eq)
   ...| (_ , _ ∷ _)   | [ eq1 ] rewrite ∷-injectiveˡ eq
      = ×-map₁ there (span-correct-1 (cong proj₂ eq1))

   span-hon : ∀ {xs dis hon : List Member} {x : Member}
            → span Meta-dishonest? xs ≡ (dis , x ∷ hon)
            → x ∈ xs × Meta-Honest-PK (getPubKey x)
   span-hon = span-correct-1 ∘ (cong proj₂)
```

Reply:
It can't be empty because we have `span Meta-dishonest? xs ≡ (dis , x ∷ hon)`, therefore `xs` must have at least `x`. 
I simplified a little, I thought we needed to prove the base case in order to do induction but we don't. Check if it's ok. But if you still prefer your solution I can change.

Regarding the warnings I am not getting them, we can check this out later.

Reply:
About the warning, you probably want to customize `agda2-highlight-catchall-clause-face` in emacs to make these warnings visible.